### PR TITLE
Emergent Sanitation Gameplay

### PR DESCRIPTION
**Pull request by ike709**
Adds everything necessary for a janitor gameplay loop (minus disposals, but that's already being worked on).

### Puddles
By default, a puddle with a volume of 20 units or less will evaporate 1 unit every 20 seconds. It will gradually become more transparent as it does. 20 units is also the maximum default volume before spillover, so essentially all puddles will eventually evaporate. I'm a bit iffy on this from a gameplay perspective.

Using a wet mop on a puddle will remove either the puddle volume, mop volume, or 5 units. Whichever is smallest.
If the mop removes the entire remaining volume of the puddle (<= 5 units), it will create a new puddle from the mop's reagents with that amount. This is to replicate SS13 wet floors.

Puddles with a volume >= 3 units (by default) are slippery. Galoshes or walking will prevent slipping.

### Trash
Added trash spawners all over the map. These have a small chance to spawn a vomit puddle or a large piece of trash, like a pizza box. Failing that, they have a high (but not guaranteed) chance to spawn a small piece of trash, such as an empty bag of chips.

### The Janitor's Trusty Tools:tm:
The janitor's closet starts with a mop, bucket, three wet floor signs, and two trashbags. Currently the only pair of galoshes are in the janitor's loadout. I'm open to adding a pair to the locker and/or moving the janitor's loadout galoshes to the locker.

### CODE REVIEW

File: Content.Server/GameObjects/Components/Fluids/PuddleComponent.cs [hunk 1]
```diff
@@ -49,7 +50,7 @@ public class PuddleComponent : Component
         public override string Name => "Puddle";
 
         private CancellationTokenSource _evaporationToken;
-        private ReagentUnit _evaporateThreshold; // How few <Solution Quantity> we can hold prior to self-destructing
+        public ReagentUnit _evaporateThreshold; // How few <Solution Quantity> we can hold prior to self-destructing
```
Comment (Zumorica) [inline Content.Server/GameObjects/Components/Fluids/PuddleComponent.cs:53]: If you're gonna make this public, encapsulate it in a property or auto-property. Also this isn't following naming standards. It should be "EvaporateThreshold". I'd personally keep this private and make a property.

File: Content.Server/GameObjects/Components/Fluids/PuddleComponent.cs [hunk 2]
```diff
@@ -174,25 +179,49 @@ public void CheckEvaporate()
...
+            UpdateAppearance();
+
+            if (CurrentVolume < ReagentUnit.New(3) && Owner.TryGetComponent(out SlipperyComponent existingSlipperyComponent))
+            {
+                Owner.RemoveComponent<SlipperyComponent>();
+            }
+            else if (CurrentVolume >= ReagentUnit.New(3) && !Owner.TryGetComponent(out SlipperyComponent newSlipperyComponent))
+            {
+                Owner.AddComponent<SlipperyComponent>();
+            }
```
Comment (Zumorica) [inline Content.Server/GameObjects/Components/Fluids/PuddleComponent.cs:209]: Maybe make this optional? I suppose that not all puddles will be slippery.

File: Content.Server/GameObjects/Components/Fluids/PuddleComponent.cs [hunk 3]
```diff
@@ -174,25 +179,49 @@ public void CheckEvaporate()
...
+            if (_evaporateThreshold == ReagentUnit.New(-1) || CurrentVolume > _evaporateThreshold)
             {
                 return;
             }
 
             _evaporationToken = new CancellationTokenSource();
 
             // KYS to evaporate
-            Timer.Spawn(TimeSpan.FromSeconds(_evaporateTime), CheckEvaporate, _evaporationToken.Token);
+            Timer.Spawn(TimeSpan.FromSeconds(_evaporateTime), Evaporate, _evaporationToken.Token);
```
Comment (Zumorica) [inline Content.Server/GameObjects/Components/Fluids/PuddleComponent.cs:219]: I mean, this works but there's a `Timer.SpawnRepeating` method for a reason.

File: Content.Server/GameObjects/Components/Markers/TrashSpawnerComponent.cs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        public float RareChance { get; set; } = 0.05f;
+
+        [ViewVariables(VVAccess.ReadWrite)]
+        public float Offset { get; set; } = 0.2f;
+
+        public override void ExposeData(ObjectSerializer serializer)
+        {
+            base.ExposeData(serializer);
+            serializer.DataField(this, x => RarePrototypes, "rare_prototypes", new List<string>());
+            serializer.DataField(this, x => RareChance, "rare_chance", 0.05f);
```
Comment (Zumorica) [inline Content.Server/GameObjects/Components/Markers/TrashSpawnerComponent.cs:45]: These names should follow the [YAML naming guidelines](https://hackmd.io/@ss14/code-quality-control).

>### Naming
>`PascalCase` is used for IDs and component names. Everything else uses `camelCase`.

File: Content.Server/GameObjects/Components/Markers/TrashSpawnerComponent.cs [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+                entity.Transform.LocalPosition += new Vector2(random.NextFloat() * Offset * x_negative, random.NextFloat() * Offset * y_negative);
+            }
+
+        }
+
+        public override void MapInit()
+        {
+            Spawn();
+            Owner.Delete();
+            return;
```
Comment (Zumorica) [inline Content.Server/GameObjects/Components/Markers/TrashSpawnerComponent.cs:84]:
```suggestion
```

File: Content.Server/GameObjects/Components/Movement/SlipperyComponent.cs [hunk 6]
```diff
@@ -90,6 +91,11 @@ public void CollideWith(IEntity collidedWith)
             if (percentage < IntersectPercentage)
                 return;
 
+            if(!EffectBlockerSystem.CanSlip(collidedWith))
+            {
+                return;
+            }
```
Comment (Zumorica) [inline Content.Server/GameObjects/Components/Movement/SlipperyComponent.cs:97]: For consistency with the code above

```suggestion
                return;
```

**Comment by Zumorica** [comment]
Apparently saltern.yml conflicts... I'm so sorry for you.

**Comment by ike709** [comment]
Puddles will now also say in their description if they're slippery.

**Pull request merged**